### PR TITLE
[Flight] Taint APIs

### PR DESCRIPTION
**Pull request by sebmarkbage**
This lets a registered object or value be "tainted", which we block from crossing the serialization boundary. It's only allowed to stay in-memory.

This is an extra layer of protection against mistakes of transferring data from a data access layer to a client. It doesn't provide perfect protection, because it doesn't trace through derived values and substrings. So it shouldn't be used as the only security layer but more layers are better.

`taintObjectReference` is for specific object instances, not any nested objects or values inside that object. It's useful to avoid specific objects from getting passed as is. It ensures that you don't accidentally leak values in a specific context. It can be for security reasons like tokens, privacy reasons like personal data or performance reasons like avoiding passing large objects over the wire.

It might be privacy violation to leak the age of a specific user, but the number itself isn't blocked in any other context. As soon as the value is extracted and passed specifically without the object, it can therefore leak.

`taintUniqueValue` is useful for high entropy values such as hashes, tokens or crypto keys that are very unique values. In that case it can be useful to taint the actual primitive values themselves. These can be encoded as a string, bigint or typed array. We don't currently check for this value in a substring or inside other typed arrays.

Since values can be created from different sources they don't just follow garbage collection. In this case an additional object must be provided that defines the life time of this value for how long it should be blocked. It can be `globalThis` for essentially forever, but that risks leaking memory for ever when you're dealing with dynamic values like reading a token from a database. So in that case the idea is that you pass the object that might end up in cache.

A request is the only thing that is expected to do any work. The principle is that you can derive values from out of a tainted
entry during a request. Including stashing it in a per request cache. What you can't do is store a derived value in a global module level cache. At least not without also tainting the object.

### CODE REVIEW

File: packages/react/src/ReactTaint.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    throw new Error('Not implemented.');
+  }
+  // eslint-disable-next-line react-internal/safe-string-coercion
+  message = '' + (message || defaultMessage);
+  if (
+    lifetime === null ||
+    (typeof lifetime !== 'object' && typeof lifetime !== 'function')
+  ) {
+    throw new Error(
+      'To taint a value, a life time must be defined by passing an object that holds ' +
```
Comment (sophiebits) [inline packages/react/src/ReactTaint.js:70]: lifetime

File: packages/react/src/ReactTaint.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    if (kind === 'object' || kind === 'function') {
+      throw new Error(
+        'taintUniqueValue cannot taint objects or functions. Try taintObjectReference instead.',
+      );
+    }
+    throw new Error(
+      'Cannot taint a ' +
+        kind +
+        ' because the value is too general and cannot be ' +
+        'a secret by',
```
Comment (sophiebits) [inline packages/react/src/ReactTaint.js:99]: typo?

**Review by sophiebits — APPROVED** [review]
Is there a meaningful reason to use the set of pending requests when the object is freed vs when taint is called?

File: packages/react-client/src/__tests__/ReactFlight-test.js [hunk 3]
```diff
@@ -1446,4 +1463,202 @@ describe('ReactFlight', () => {
...
+  // @gate enableTaint
+  it('errors when a tainted string is serialized', async () => {
+    function UserClient({user}) {
+      return <span>{user.name}</span>;
+    }
+    const User = clientReference(UserClient);
+
+    const process = {
+      env: {
+        SECRET: '3e971ecc1485fe78625598bf9b6f85db',
```
[the remark this answers is not in the document]
Comment (sebmarkbage) [reply packages/react-client/src/__tests__/ReactFlight-test.js:1525]: I was wondering who would be the first nerd to figure this out. Did you notice the hidden pattern for the other ones tho?

[the remark this answers is not in the document]
Comment (sophiebits) [reply packages/react-client/src/__tests__/ReactFlight-test.js:1525]: no!

**Pull request merged**